cookbook/case_studies/feature_engineering/feast_integration/feast_wf.py

```diff
@@ -0,0 +1,351 @@
+            Feature(name="nasogastric reflux PH", dtype=ValueType.FLOAT),
+        ],
+        input=FileSource(
+            path=str(parquet_file),
+            event_timestamp_column="timestamp",
+        ),
+        ttl=timedelta(days=1),
+    )
+
+    fs.apply([horse_colic_entity, horse_colic_feature_view])
+    return repo_path, horse_colic_feature_view.name
+
+
+@task
+def retrieve_offline(repo_path: str) -> pd.DataFrame:
+    fs = FeatureStore(repo_path=repo_path)
+    entity_df = pd.DataFrame.from_dict(
+        {
+            "Hospital Number": [
+                "530101",
+                "5290409",
+                "5291329",
+                "530051",
+                "529518",
+                "530101",
+                "529340",
+                "5290409",
+                "530034",
+            ],
+            "event_timestamp": [
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 7, 5, 11, 36, 1),
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 7, 5, 11, 50, 40),
+                datetime(2021, 6, 25, 16, 36, 27),
+            ],
+        }
+    )
+
+    retrieval_job = fs.get_historical_features(
+        entity_df=entity_df,
+        feature_refs=FEAST_FEATURES,
+    )
+
+    feature_data = retrieval_job.to_df()
+    return feature_data
+
+
+# %%
+# Next, we train the Naive Bayes model using the data that's been fetched from the feature store.
+@task
+def train_model(
+    dataset: pd.DataFrame, data_class: str, feature_view_name: str
+) -> JoblibSerializedFile:
+    X_train, _, y_train, _ = train_test_split(
+        dataset,
+        dataset[feature_view_name + "__" + data_class],
+        test_size=0.33,
+        random_state=42,
+    )
+    model = GaussianNB()
+    model.fit(X_train, y_train)
+    model.feature_names = list(X_train.columns.values)
+    fname = "model.joblib.dat"
+    joblib.dump(model, fname)
+    return fname
+
+
+# %%
+# To perform inferencing, we define two tasks: ``store_online`` and ``retrieve_online``.
+#
+# .. list-table:: Decoding the ``Feast`` Jargon
+#    :widths: 25 25
+#
+#    * - ``materialize()``
+#      - Materialize data from the offline store into the online store.
+#    * - ``get_online_features()``
+#      - Retrieves the latest online feature data.
+#
+# .. note::
+#   One key difference between the online store and data source is that only the latest feature values are stored per entity key. No historical values are stored.
+#   Our dataset has two such entries with the same ``Hospital Number`` but different time stamps. Only data point with the latest timestamp is picked from the online store.
+@task
+def store_online(repo_path: str) -> str:
+    store = FeatureStore(repo_path=repo_path)
+    store.materialize(
```

Comment:
why is the time hard coded?

Reply:
We can take inputs from the user. But then, it can be days, hours, minutes, ... etc. We'll have to ask the user to give four inputs: two specifying the start and end time format, and the other two specifying their respective values.